File: scripts/glm.py
#!/usr/bin/env python3
"""Call GLM-5 via ZAI OpenAI-compatible API.
Reads the API key from OpenClaw's auth-profiles.json.
Usage: echo "prompt" | python3 glm.py
   or: python3 glm.py "prompt text"
"""
import sys
import json
import urllib.request
import urllib.error
import os
from pathlib import Path

BASE_URL = "https://api.z.ai/api/coding/paas/v4"
MODEL = "glm-5"

def get_key() -> str:
    # 1. Env override
    key = os.environ.get("ZAI_API_KEY") or os.environ.get("GLM_API_KEY")
    if key:
        return key

    # 2. OpenClaw auth-profiles.json (main agent — has the full key store)
    paths = [
        Path.home() / ".openclaw/agents/main/agent/auth-profiles.json",
        Path.home() / ".openclaw/agents/builder/agent/auth-profiles.json",
    ]
    for p in paths:
        if p.exists():
            data = json.loads(p.read_text())
            profile = data.get("profiles", {}).get("zai:default", {})
            k = profile.get("key", "")
            if k:
                return k

    raise RuntimeError("ZAI API key not found. Set ZAI_API_KEY env var or ensure ~/.openclaw/agents/main/agent/auth-profiles.json has zai:default profile.")

def call(prompt: str) -> str:
    key = get_key()
    payload = json.dumps({
        "model": MODEL,
        "messages": [{"role": "user", "content": prompt}],
        "max_tokens": 8192,
        "temperature": 0.7,
    }).encode()

    req = urllib.request.Request(
        f"{BASE_URL}/chat/completions",
        data=payload,
        headers={
            "Authorization": f"Bearer {key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            data = json.loads(resp.read())
            return data["choices"][0]["message"]["content"]
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        raise RuntimeError(f"GLM-5 API error {e.code}: {body[:500]}")

def main():
    if len(sys.argv) > 1:
        prompt = " ".join(sys.argv[1:])
    else:
        prompt = sys.stdin.read()

    if not prompt.strip():
        print("[GLM-5: empty prompt]", file=sys.stderr)
        sys.exit(1)

    result = call(prompt)
    print(result)

if __name__ == "__main__":
    main()

File: internal/models/glm.go
package models

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"time"
)

type GLMProvider struct{}

func (g *GLMProvider) ID() string   { return "glm5" }
func (g *GLMProvider) Name() string { return "GLM-5 (ZhipuAI)" }

func (g *GLMProvider) Available(ctx context.Context) bool {
	_, err := g.getAPIKey()
	return err == nil
}

func (g *GLMProvider) Plan(ctx context.Context, prompt string, timeout time.Duration) (string, error) {
	text, _, _, err := g.PlanWithTokens(ctx, prompt, timeout)
	return text, err
}

func (g *GLMProvider) PlanWithTokens(ctx context.Context, prompt string, timeout time.Duration) (string, int, int, error) {
	apiKey, err := g.getAPIKey()
	if err != nil {
		return "", 0, 0, err
	}

	payload := map[string]interface{}{
		"model": "glm-5",
		"messages": []map[string]string{
			{"role": "user", "content": prompt},
		},
		"max_tokens":  8192,
		"temperature": 0.7,
	}

	data, err := json.Marshal(payload)
	if err != nil {
		return "", 0, 0, err
	}

	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, "POST", "https://api.z.ai/api/coding/paas/v4/chat/completions", bytes.NewReader(data))
	if err != nil {
		return "", 0, 0, err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", 0, 0, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", 0, 0, fmt.Errorf("GLM-5 API error: %s - %s", resp.Status, string(body))
	}

	var result struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
		Usage struct {
			PromptTokens     int `json:"prompt_tokens"`
			CompletionTokens int `json:"completion_tokens"`
		} `json:"usage"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", 0, 0, err
	}

	if len(result.Choices) == 0 {
		return "", 0, 0, fmt.Errorf("no content in response")
	}

	return result.Choices[0].Message.Content, result.Usage.PromptTokens, result.Usage.CompletionTokens, nil
}

func (g *GLMProvider) getAPIKey() (string, error) {
	if key := os.Getenv("ZAI_API_KEY"); key != "" {
		return key, nil
	}
	if key := os.Getenv("GLM_API_KEY"); key != "" {
		return key, nil
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("could not get home directory")
	}

	paths := []string{
		filepath.Join(home, ".openclaw/agents/main/agent/auth-profiles.json"),
		filepath.Join(home, ".openclaw/agents/builder/agent/auth-profiles.json"),
	}

	for _, path := range paths {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}

		var authProfiles struct {
			Profiles map[string]struct {
				Key string `json:"key"`
			} `json:"profiles"`
		}

		if err := json.Unmarshal(data, &authProfiles); err != nil {
			continue
		}

		if profile, ok := authProfiles.Profiles["zai:default"]; ok && profile.Key != "" {
			return profile.Key, nil
		}
	}

	return "", fmt.Errorf("ZAI API key not found. Set ZAI_API_KEY env var or ensure ~/.openclaw/agents/main/agent/auth-profiles.json has zai:default profile")
}
